### ISSUE TITLE
url expression in pyparsing_common

### ISSUE THREAD

**Issue by WolfgangFahl** [pyparsing/pyparsing#249]
see e.g. 
* https://mathiasbynens.be/demo/url-regex
* https://gist.github.com/dperini/729294 
here is  code snippet:

```python
@staticmethod    
    def getUriRegexp():
        '''
        get a regular expression for an URI
        '''
        # https://mathiasbynens.be/demo/url-regex
        # https://gist.github.com/dperini/729294 
        uriRegexp=(
        r"^"
        # protocol identifier
        r"(?:(?:(?:https?|ftp):)?//)"
        # user:pass authentication
        r"(?:\S+(?::\S*)?@)?"
        r"(?:"
        # IP address exclusion
        # private & local networks
        r"(?!(?:10|127)(?:\.\d{1,3}){3})"
        r"(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})"
        r"(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})"
        # IP address dotted notation octets
        # excludes loopback network 0.0.0.0
        # excludes reserved space >= 224.0.0.0
        # excludes network & broadcast addresses
        # (first & last IP address of each class)
        r"(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])"
        r"(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}"
        r"(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))"
        r"|"
        # host & domain names, may end with dot
        # can be replaced by a shortest alternative
        # r"(?![-_])(?:[-\w\u00a1-\uffff]{0,63}[^-_]\.)+"
        # r"(?:(?:[a-z\u00a1-\uffff0-9]-?)*[a-z\u00a1-\uffff0-9]+)"
        # # domain name
        # r"(?:\.(?:[a-z\u00a1-\uffff0-9]-?)*[a-z\u00a1-\uffff0-9]+)*"
        r"(?:"
        r"(?:"
        r"[a-z0-9\u00a1-\uffff]"
        r"[a-z0-9\u00a1-\uffff_-]{0,62}"
        r")?"
        r"[a-z0-9\u00a1-\uffff]\."
        r")+"
        # TLD identifier name, may end with dot
        r"(?:[a-z\u00a1-\uffff]{2,}\.?)"
        r")"
        # port number (optional)
        r"(?::\d{2,5})?"
        # resource path (optional)
        r"(?:[/?#]\S*)?"
        r"$")
        compiled=re.compile(uriRegexp,re.RegexFlag.I |re.RegexFlag.UNICODE)
        return compiled
```

**Comment by ptmcg** [comment]
Thanks for submitting this. That link also includes a good set of expected good and expected fail URLs.

(I think emails are even more difficult to parse with a general expression.)

**Comment by WolfgangFahl** [comment]
py-sidif has an implementation which i'd love to see in the base library
https://github.com/WolfgangFahl/py-sidif/blob/e49fe64d9a330871aec386d25f70a21ade51f27c/sidif/sidif.py#L154
Some of the examples made it into the unit test
https://github.com/WolfgangFahl/py-sidif/blob/e49fe64d9a330871aec386d25f70a21ade51f27c/tests/test_sidif.py#L45
see
https://github.com/WolfgangFahl/py-sidif/blob/e49fe64d9a330871aec386d25f70a21ade51f27c/tests/test_sidif.py#L22

**Comment by ptmcg** [comment]
So sorry for dragging this out. I was originally reluctant to include this because this is sort of a massive regex to add to pyparsing_common, but after looking at it in detail, it hangs pretty well together - the comments help a lot!

If you would open a pull request for adding a url to pyparsing_common, I'll try to get it in to 3.0.0 still.  If you can, please add group names for scheme, auth, host, port, and path, and wrap in a pyparsing Regex. Any group names that you define in your re string will be preserved as results names when parsed within a Regex. If you are not familiar with group names, that's fine, I can add them after merging your PR.

**Comment by ptmcg** [comment]
Well, this turned out to be simpler than I thought, including adding the group names. I'll drop it into the next release.

**Comment by WolfgangFahl** [comment]
Thx for being willing and able to finish this on your own.  I am currently occupied with a dear family member having passed away :-(

**Comment by ptmcg** [comment]
Released in 3.0.0rc2
